src/codegen/util/sorter.cpp

```diff
@@ -6,167 +6,271 @@
-  return curr_pos_ == rhs.curr_pos_;
-}
+  timer.Stop();
+  LOG_DEBUG("Total sort time: %.2lf ms", timer.GetDuration());
+  timer.Reset();
+  timer.Start();
 
-bool Sorter::Iterator::operator!=(const Sorter::Iterator &rhs) const {
-  return !(*this == rhs);
-}
+  // Where the merging work units are collected
+  struct MergeWork {
+    std::vector<std::pair<char **, char **>> input_ranges;
+    char **destination = nullptr;
+    MergeWork(std::vector<std::pair<char **, char **>> &&inputs, char **dest)
+        : input_ranges(std::move(inputs)), destination(dest) {}
+  };
+  std::vector<MergeWork> merge_work;
+
+  // Compute global separators from locally-computed separators
+  auto cmp = [this](char *l, char *r) { return cmp_func_(l, r) < 0; };
+  char **write_pos = tuples_.data();
+  std::vector<char **> last_pos(sorters.size());
+  for (uint32_t idx = 0; idx < separators.size(); idx++) {
+    // Sort the local separators and choose the median
+    char *separator = nullptr;
+    if (idx < separators.size() - 1) {
+      std::sort(separators[idx].begin(), separators[idx].end(), cmp);
+      separator = separators[idx][sorters.size() / 2];
+    }
+
+    std::vector<std::pair<char **, char **>> input_ranges;
+    uint64_t part_size = 0;
+    for (uint32_t sort_idx = 0; sort_idx < sorters.size(); sort_idx++) {
+      Sorter *sorter = sorters[sort_idx];
+      char **start = (idx == 0 ? sorter->tuples_.data() : last_pos[sort_idx]);
+      char **end = sorter->tuples_.data() + sorter->tuples_.size();
+      if (idx < separators.size() - 1) {
+        end = std::upper_bound(start, end, separator, cmp);
+      }
+      if (start != end) {
+        input_ranges.emplace_back(start, end);
+      }
+      part_size += (end - start);
+      last_pos[sort_idx] = end;
+    }
+    // Add work
+    merge_work.emplace_back(std::move(input_ranges), write_pos);
+    // Bump new write position
+    write_pos += part_size;
+  }
+
+  timer.Stop();
+  LOG_DEBUG("Work generation time: %.2lf ms", timer.GetDuration());
+  timer.Reset();
+  timer.Start();
+
+  common::synchronization::CountDownLatch merge_latch{merge_work.size()};
+
+  // Divide up the merging work across all threads
+  auto heap_cmp = [this](const std::pair<char **, char **> &l,
+                         const std::pair<char **, char **> &r) {
+    return !(cmp_func_(*l.first, *r.first) < 0);
+  };
+  for (auto &work : merge_work) {
+    work_pool.SubmitTask([&work, &merge_latch, &heap_cmp, &cmp] {
+      // Heap
+      std::priority_queue<std::pair<char **, char **>,
+                          std::vector<std::pair<char **, char **>>,
+                          decltype(heap_cmp)> heap(heap_cmp, work.input_ranges);
+      char **dest = work.destination;
+      while (!heap.empty()) {
+        auto top = heap.top();
+        heap.pop();
+        *dest++ = *top.first;
+        if (top.first + 1 != top.second) {
+          heap.emplace(top.first + 1, top.second);
+        }
```

Comment:
Sorry for the many questions on this, I just want to make sure I get this correct. I can also come by in person if that's easier. 

Correct me if I am wrong:
* in a parallel pipeline, the control flow joins in the main execution thread before sorting
* then new tasks are spawned for the actual sorting, where every thread has a sorter instance that sorts its data locally (line 140 [1])
* then the median-of-medians is calculated by picking values from same indexes in the single sorter instances and taking the mean per index (done in main execution thread)
* with this median the sorted data per instance if partitioned and put into merge packages (also main execution thread)
* finally, new tasks are spawned with the actual merging:
  * the priority queue holds a set of to-merge sorted partitions (the one with smallest first value first)
  * the first value from the first partition is written to the destination buffer and the partition with all remaining values is put back into the queue
  * while merging the data is read from the thread-local sorter instances' memory and written into the main sorter instance's memory

[1] https://github.com/cmu-db/peloton/pull/1304/files#diff-6e82859adc9b308f36bda0aab4a0fbfeR140

Reply:
Correct. I've added more comments to help.